### ISSUE TITLE
Implement a JSDoc @import tag

### ISSUE THREAD

**Issue by DanielRosenwasser** [microsoft/TypeScript#22160]
# Background

#22158 tracks referencing types from a given module using JSDoc-style namepaths. Given that the syntax is somewhat unintuitive and predates the concept of ECMAScript modules, we would like to support a more ergonomic form if we feel it would be helpful.

# Options

Bikeshedding time. 🚲 🏠 

## `@from` `@import`

```js
/**
 * @from "express"
 * @import { Request, Response as CoolResponse } from
 * @import Default
 * @import * as ns
 */
```

### Pros

* Allowing `@from` to precede named import clauses would be easier for completions (this is one of our oldest-duped requests #2371)
* Can have multiple `import` clauses
* Could hypothetically support `@from` before `@import` and vice versa.

### Cons

* Doesn't totally look like ESModule syntax which is just extra cognitive overhead.

## ECMAScript Import-based

```js
/**
 * @import { x, y as z, default as Default } from "express"
 */
```

### Pros

* Nothing to learn if you already know ECMAScript syntax.

### Cons

* Less optimal for completions

# Issues with the Above

The options above don't make it explicit that only types are being imported. We could play around with keyword/tag placement (e.g. `@importtype`, `@import type`, etc.)

**Comment by aozgaa** [comment]
This was previously discussed in https://github.com/Microsoft/TypeScript/issues/14377

**Comment by DanielRosenwasser** [comment]
We're going to hold off on any new JSDoc syntax, and wait for more feedback here. As per #22445, our current recommendation is to use

```js
/**
 * @typedef {import("express")} express
 */
```

**Comment by mhegazy** [comment]
closing in favor of  #14844

**Comment by jkrems** [comment]
@DanielRosenwasser Did the syntax for module namespaces change? I tried scanning through the linked issues but since none of them are showing the JSDoc equivalent, it's somewhat hard to follow.

```js
/**
 * The following works using latest typescript@next (2.9.0-dev.20180506):
 *
 * @typedef {import('http').IncomingMessage} IncomingMessage
 * @typedef {import('http').ServerResponse} ServerResponse
 *
 * But this fails:
 *
 * > Module 'http' does not refer to a type, but is used as a type here.
 *
 * @typedef {import('http')} http
 */
```

**Comment by nojvek** [comment]
Hi Typescript Magic Elves,

Do you think you'll be able to get any decisions on what the final DRY-er jsdoc import syntax will look like?

Just curious what the blocking items are?

**Comment by nojvek** [comment]
I hope this doesn’t get stuck in analysis paralysis fever like some of the other issues. 

I am Typescript js-doccing a medium sized project and this is a pretty painful experience. It feels like a very subpar experience to TS.

**Comment by nojvek** [comment]
Kind ping to typescript authors and @DanielRosenwasser . Has any decision been regarding this?

Using tsc as a typechecker for js with a verbose syntax like this adds up quite a bit of friction.

**Comment by nojvek** [comment]
Been many months on this now. Is there any progress on this issue? I'd really love to see `/** @import { x, y as z, default as Default } from "express" */`

We're adding jsdoc types to half a million line codebase and we feel the pain of the verbose syntax every single day.

Please please please please please make a decision on this. I'd be happy to work on a PR.

**Comment by s-bauer** [comment]
Any progress on this? This is a quite important feature!!

**Comment by Jemt** [comment]
I like the way we can do it with TypeScript now. I'm already using it a _lot_.

```
// Inline approach
function(/** @type {import("./../Interfaces").Context} */ context)
{
    // Code goes here..
}

// Re-usable type definition
/** @typedef {import("./../Interfaces").Configuration} Config */
```

**Comment by jeffersoneagley** [comment]
Just as an FYI on another couple approaches I use: 
Assuming some package my-npm-package
Approach 1 in regular JS code: 
```
import * as MyNpmPackage from 'my-npm-package'

/** 
 * @type {MyNpmPackage.ISomeType} mySomeType
 */
const mySomeType = etc..
```
Method I use on one project:
I actually use Typescript as a compiler/linter/etc and have it set to allowJs:true, checkJs:true among a big list of other things (basically all the features of TS but I don't have to write directly in the language). As you might guess, this allows me to just do:
```
import {ISomeType} from 'my-npm-package';
```
without needing to commit to migrating a whole codebase to actually using typescript.

**Comment by orzechowskid** [comment]
@jeffersoneagley I would love to see how you've acccomplished your approach #2 . I'm attempting to do the exact same thing with a project of mine but I'm failing to get my IDE to recognize any types imported from external files.

**Comment by rjgotten** [comment]
> I like the way we can do it with TypeScript now.

On the contrary; I _hate_ the way you have to do it now.

It's extremely verbose, which breaks down into an unreadable mess quickly. E.g. when you have to import a generic type and also have to import its type constraints. E.g.
```js
/**
 * @param {value:import('./other/a').GenericA<import('./other/b').ConstraintB, import('./other/c').ConstraintC>}
 */
``` 

To avoid creating that kind of mess, you have to import types up front and create local aliases for them with `@typedef` instructions. If you need multiple types you cannot bulk import them. You need to do it one by one and alias each one with its own `@typedef` inflating the JS file with a lot of noise.

You'd think you'd be able to use namespaces, but that's not supported properly. At the very least it's buggy. I tried aforementioned `typeof` approach and it didn't work for me.

Just being able to put a single comment at the top of a JS file, before or after the block of ES6 imports, with a conceptual `@import` JSDoc equivalent to import _types_ seems a major; **MAJOR** improvement.

It would probably also solve one other big pain point: Currently if you author JS with split `.d.ts` definition files for typing said JS file itself _is not typed_ with its own `.d.ts` file. All other files that import the JS apply the `.d.ts` typings, but _not_ the file itself.

With a proper import styled after ES6 you could do something like:
```js
/** @import * from "./self" */
```
And have the file's declared typings applied to itself that way.

**Comment by nojvek** [comment]
It's been more than 2 years. @DanielRosenwasser is there any progress on this ?

**Comment by kfdf** [comment]
I'm new to javascript/typescript and might be misunderstanding something, but doesn't this work?
```
// lib.js
/**
@typedef {object} Person
@property {string} name 
@property {number} age */

/**
@template T
@typedef {object} Node 
@property {T} value
@property {Node<T>} next */

/** @type {Person} */
export const Person = null;

/** 
@template T
@type {Node<T>} */
export const Node = null;
```
And then
```
import { Person, Node } from './lib.js';

/**
@param {Node<Person>} head */
function printList(head) {
  if (!head) return;
  let { name, age } = head.value;
  console.log(name, age);
  printList(head.next);
}
printList({ 
  value: { age: 5, name: 'John' },
  next: {
    value: { age: 10, name: 'Ann' },
    next: null
  }
});
```
And I got all the red squiggles and intellisence when typing that.
Sure it does spill some bits of typescript declarations into your javascript but other than that it seems like relatively smooth experience...

**Comment by ExE-Boss** [comment]
@kfdf It’s ugly and may cause unwanted circular dependencies, which can cause breakages at runtime because of unexpected module execution order.

**Comment by rjgotten** [comment]
> @kfdf 
> but doesn't this work?

Completely different problem-space.

There are TypeScript (and VS Code JS intellisense) constraints you cannot express with typedef which _need_ `.d.ts` files. And it's importing 'definitions only' from those `.d.ts` files into JS files which is the problem.

**Comment by bmeck** [comment]
Just want to note the workaround in https://github.com/microsoft/TypeScript/issues/22160#issuecomment-371930983 seems to drop the documentation of the type:

```js
/**
 * This is documented
 * @typedef {{foo:'bar'}} Foo
 */
```

```js
/**
 * @typedef {import('foo').Foo} Foo
 */
// Foo does not have a description
```

**Comment by ExE-Boss** [comment]
Also, <https://github.com/microsoft/TypeScript/issues/22160#issuecomment-371930983> doesn’t support optional type parameters.

**Comment by jeffersoneagley** [comment]
Any updates on this? I'm still importing things at the top of the file, but CRA of course complains in the preflight.

**Comment by noinkling** [comment]
I've resorted to adding `export as namespace foo` to my .d.ts modules. It makes the namespace globally available which isn't ideal, but it's the only way I can find to avoid repeated `import(...)` spam, since there doesn't seem to be any other way to alias a namespace in JSDoc (which I've previously commented on in another issue: https://github.com/microsoft/TypeScript/issues/14233#issuecomment-479422095).

**Comment by RobinBlomberg** [comment]
@jeffersoneagley I had great success with approach #1, and I found a way to make it interoperable with normal .js/.ts files inside ESM modules. This approach takes advantage of how TypeScript appears to merge file extensions if the filenames are the same.

**my-module.ts**

```typescript
export type FooString = 'foo' | 'bar' | 'baz';

export type FooObject = { foo: FooString; };
```

**my-module.js**

```javascript
/**
 * @file Dummy export in order to avoid a runtime ERR_MODULE_NOT_FOUND error.
 */

export {};
```

**index.js**

```javascript
import * as MyModule from './my-module.js';

/**
 * @param {MyModule.FooString} fooString
 * @return {MyModule.FooObject}
 */
export const createFooObject = (fooString) => {
  return { foo: fooString };
};
```

Now it seems a bit silly to have to create a dummy export file every time, but in my case, I need to define and import 100+ namespaced types, and I really haven't found any other reasonable way to do it.

**Comment by nanxiaobei** [comment]
Hope to use JSDoc syntax to import typescript definition to *.js file.

**Comment by jeffersoneagley** [comment]
@RobinBlomberg I actually define mine in types.d.ts files around the app now and they basically give every js file in the directory the ability to access types defined in pure TS (don't use any import statements other than inline though, top-of-file imports will turn the type file into a module and then you can't easily import it into js)

**Comment by RobinBlomberg** [comment]
@jeffersoneagley Genius! I'll have to try this.

**Comment by NemoStein** [comment]
It seems that this issue will never reach a proper conclusion, but this is blocking #46011.
We need a way of opting out of `@typedef` auto exporting the type!

`/** @import Type from './path/to/module.js' */` would be, IMO, the perfect solution, but something else is needed if this isn't possible.

**Comment by jespertheend** [comment]
These are snippets of some of my real world code 🥲
<img width="537" alt="image" src="https://user-images.githubusercontent.com/2737650/214722408-e3beb6d1-d712-44ae-9bc0-74c145cc12d9.png">
<img width="1339" alt="image" src="https://user-images.githubusercontent.com/2737650/214722548-3f599e4e-8c8b-44fc-9660-dfe0d90262da.png">
<img width="1121" alt="image" src="https://user-images.githubusercontent.com/2737650/214722776-110a51a3-32c9-4499-9757-a724bbe5f5a3.png">

What can I say, it is what it is. I still prefer this rather than having to deal with a build step and import maps though.

**Comment by DanielRosenwasser** [comment]
Thanks to @a-tarasyuk, this should be in the next nightly release and in TypeScript 5.5. The syntax is based on ECMAScript imports:

```js
/**
 * @import * as foo from "some-module-path"
 */

/**
 * @import { x, y as z, default as Default } from "another-module-path"
 */
```

**Comment by waynesbrain** [comment]
I can't get `@import` OR `@typedef {import()}` working in my `.ts` files here in my project which I just upgraded to TS 5.6.3 in the `next` branch - https://github.com/jrfso/jrfs/blob/next/packages/core/src/types.ts#L6

If I create a `.js` though, this works `/** @import { FileTree } from "@/FileTree" */` and I've tried all variations in `.ts` files including `from "./FileTree"`.

**Comment by waynesbrain** [comment]
@turadg Any advice? Am I doing it wrong? (Sorry!)

**Comment by jespertheend** [comment]
@waynesbrain I believe the `@import` tag only works in `.js` files. In `.ts` files you can use TypeScript syntax like so:

```ts
import type { FileTree } from "@/FileTree";
```

**Comment by waynesbrain** [comment]
Thanks @jespertheend - I see now that the docs aren’t explicit that you can only do this in a JavaScript file, but it’s the only type of file that’s mentioned.

I wish they would have implemented it in TS because if I import this type just for documentation using the TS syntax then I get errors for having an unused import.
